andela/ah-backend-prime
# #164798291 Set Up social authentication for user registration

**What does this PR do?**
    Sets up Social authentication for user registration

**Description of Task to be completed?**

- Set up requests to Google, Twitter and Facebook services for user token

- Decode User Token

- Register user using the information obtained from the decoded token

**How should this be manually tested?**

- After cloning the repo, cd into it and run the following commands to activate the virtual environment
   `virtualenv venv`
  ` source venv/bin/activate`

- Checkout the branch with the validation implementation
  `git checkout ft-social-login-164798291`

- Install all requirements
   `pip install -r requirements.txt`

- Source the environment variables
  ` source .env`

- Run the migrations
   `python manage.py makemigrations`
   `python manage.py migrate`

- Run the server
   `python manage.py runserver`

**Testing Facebook Social Authentication**

- Navigate [here](https://developers.facebook.com/) to create an app

- Navigate [here](https://developers.facebook.com/tools/explorer) to obtain an access token for your account

- Navigate to Postman and use the URL `http://127.0.0.1:8000/api/social/auth/facebook/` and in the body section, paste the access token, change to a **POST** request and click **SEND** to view the response as shown in Figure 1.

<img width="1219" alt="Screen Shot 2019-04-07 at 16 18 23" src="https://user-images.githubusercontent.com/26630001/55684513-225a9f00-5954-11e9-9d4f-15ec121e3704.png">
Figure 1


**Testing Google Social Authentication**

- Navigate [here](https://developers.google.com/oauthplayground/) to obtain a google token

- In the scope selection, search for **Google OAuth2 API v2** and check all of its three fields.

- Click the Authorize APIs button 

- An Authorization code would be generated, click the **Exchange authorization code for tokens** button for an id token to be generated.

- Copy that id token, navigate to Postman, in the URL section,  insert`http://127.0.0.1:8000/api/social/auth/google/` and in the body section, paste the id token, change to a **POST** request and click **SEND** to view the response as shown in Figure 2.

<img width="1230" alt="Screen Shot 2019-04-07 at 16 34 22" src="https://user-images.githubusercontent.com/26630001/55684475-7d3fc680-5953-11e9-91aa-debf317f4e52.png">
Figure 2

**Testing Twitter Social Authentication**

- Navigate [here](https://developer.twitter.com/en/apps) to create an app on twitter

- After creating the app, navigate to the **keys and tokens** tab where you will view your consumer keys and Access keys. Get a copy of your access keys.
<img width="1137" alt="Screen Shot 2019-04-07 at 16 46 17" src="https://user-images.githubusercontent.com/26630001/55684675-a31a9a80-5956-11e9-9bec-a711e3dc6039.png">


- Navigate to Postman and in the URL section, input the URL `http://127.0.0.1:8000/api/social/auth/twitter/`, in the body section paste both the access token and the access secret token with a space separating them, change to a **POST** request and click **SEND** to view the response as shown in Figure 3.

<img width="1222" alt="Screen Shot 2019-04-07 at 16 54 56" src="https://user-images.githubusercontent.com/26630001/55684632-1b349080-5956-11e9-9a50-9f6d4029790d.png">
Figure 3

**What are the relevant pivotal tracker stories?**
[#164798291](https://www.pivotaltracker.com/story/show/164798291)
